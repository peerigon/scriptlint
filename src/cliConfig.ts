import commander from "commander";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require("../package.json");

type CliConfig = {
	packageFile?: string;
	fix?: boolean;
	strict?: boolean;
	json?: boolean;
	config?: boolean;
};

export default (argv: Array<string>): CliConfig => {
	const program = new commander.Command();
	let packageFile;

	program
		.version(`${version}`)
		.arguments("[packageFile]")
		.action((arg: string) => {
			packageFile = arg;
		})
		.option("-s, --strict", "strict mode")
		.option("-j, --json", "JSON output")
		.option("-c, --config", "inspect the config")
		.option("-f, --fix", "autofixing");

	program.parse(argv);

	// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
	const cliConfig: CliConfig = { packageFile: packageFile ?? process.cwd() };

	if (program.fix !== undefined) {
		cliConfig.fix = program.fix;
	}

	if (program.strict !== undefined) {
		cliConfig.strict = program.strict;
	}

	if (program.json !== undefined) {
		cliConfig.json = program.json;
	}

	if (program.config !== undefined) {
		cliConfig.config = program.config;
	}

	return cliConfig;
};
