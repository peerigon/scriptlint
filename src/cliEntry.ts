import cliRun from "./cli";

cliRun(process.argv);
