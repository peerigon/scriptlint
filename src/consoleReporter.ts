import chalk from "chalk";
import { Values, MessageBuffer, MessageType } from "./types";
import { makeMessage } from "./utils";

const PREFIX = "⧙։⧘";
let stashed: MessageBuffer = [];

const stash = (message: string, type: MessageType): void => {
	stashed.push({
		message,
		type
	});
};

export const warning = (template: string, values?: Values): void => {
	const message = makeMessage(template, values);

	stash(message, "warning");
};

export const success = (template: string, values?: Values): void => {
	const message = makeMessage(template, values);

	stash(message, "success");
};

export const dump = (jsonOutput: boolean): number => {
	const problemCount = stashed.length;

	if (jsonOutput) {
		// eslint-disable-next-line no-console
		console.log(JSON.stringify(stashed, null, "\t"));
	} else {
		stashed.forEach(({ message, type }) => {
			print(type, message);
		});
	}

	stashed = [];

	return problemCount;
};

export const error = (message: string): void => {
	print("error", message);
};

const print = (type: MessageType, message: string) => {
	switch (type) {
	case "error": {
		// eslint-disable-next-line no-console
		console.log(chalk.bold.red(`${PREFIX} [error] ${message}`));
		break;
	}

	case "warning": {
		const notUnderlined = chalk.yellow.bold(`${PREFIX} [${type}]`);
		const underlined = chalk.yellow.underline(message);

		// eslint-disable-next-line no-console
		console.log(`${notUnderlined} ${underlined}`);
		break;
	}
	case "success": {
		// eslint-disable-next-line no-console
		console.log(chalk.bold.gray(`${PREFIX} [✔️] ${message}`));
		break;
	}
	}
};
