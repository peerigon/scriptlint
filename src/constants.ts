import {Config} from "./types";

export const DEFAULT_CONFIG: Config = {
	strict: false,
	fix: false,
	json: true,
	config: false,
	packageFile: undefined,
	rules: {},
	customRules: [],
	ignoreScripts: [],
};

export const PROJECT_NAME = "scriptlint";

export const NAME_REGEX = /^[\d:A-Za-z]+$/;

export const NAMESPACES = [
	"build",
	"dev",
	"format",
	"other",
	"report",
	"setup",
	"start",
	"test"
];

export const DEFAULT_NPM_HOOKS = [
	"install",
	"postinstall",
	"postpack",
	"postpublish",
	"postrestart",
	"postshrinkwrap",
	"poststart",
	"poststop",
	"posttest",
	"postuninstall",
	"postversion",
	"preinstall",
	"precommit",
	"prepack",
	"prepare",
	"prepublish",
	"prepublishOnly",
	"prerestart",
	"preshrinkwrap",
	"prestart",
	"prestop",
	"pretest",
	"preuninstall",
	"preversion",
	"publish",
	"release",
	"restart",
	"shrinkwrap",
	"stop",
	"uninstall",
	"version"
];
