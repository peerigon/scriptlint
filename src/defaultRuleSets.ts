import defaultRules from "./rules";
import { Rule } from "./types";

export const optionalRules = ["natural-order"];

const strict = defaultRules
	.map((r: Rule) => r.name)
	.filter((r) => !optionalRules.includes(r));

const ruleSets = {
	default: [
		"mandatory-test",
		"mandatory-start",
		"mandatory-dev",
		"no-default-test",
	],
	strict,
	optionalRules,
};

export default ruleSets;
