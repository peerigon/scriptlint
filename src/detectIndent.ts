export type Indent = {
	type: "tab" | "space" | undefined;
	amount: number;
	indent: string;
};

export const detectIndent = (fileContents: string) => {
	const lineBeginnings = fileContents.split("\n").map((line) => {
		const noSpace = line.trimStart();
		const onlySpace = line.replace(noSpace, "");

		return onlySpace;
	});

	const indentsOnly = lineBeginnings
		.reduce<Array<string>>((acc, line, index, arr) => {
			if (index < 1) return acc;
			const lastLine = arr[index - 1];

			if (lastLine.length < line.length) {
				acc.push(line);
			}

			return acc;
		}, [])
		.sort((a, b) => {
			return a.length - b.length;
		});

	if (indentsOnly.length === 0) {
		return { type: undefined, amount: 0, indent: "" };
	}

	const indent = indentsOnly[0];
	const firstChar = indent.substring(0, 1);

	const type: Indent["type"] = (() => {
		if (firstChar === " ") {
			return "space";
		}
		if (firstChar === "\t") {
			return "tab";
		}

		return undefined;
	})();

	const amount = indent.length;

	return { type, indent, amount };
};
