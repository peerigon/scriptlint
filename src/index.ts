/*
路---------路
| {scr:pt |
|   l:nt} |
|         |
|         |
路---------路
*/

import module from "./module";

export = module;
