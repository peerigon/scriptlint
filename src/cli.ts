#!/usr/bin/env node
import loadUserConfig from "./userConfig";
import loadCliConfig from "./cliConfig";
import { makePackageFilePath } from "./utils";
import userPackageScriptContext from "./userPackageScripts";
import { loadRulesFromRuleConfig } from "./loadRules";
import execute from "./execute";
import { success, warning, dump, error } from "./consoleReporter";
import {
	DEFAULT_CONFIG,
	PROCESS_EXIT_ERROR,
	PROCESS_EXIT_OK
} from "./constants";

/* istanbul ignore next */
const processExit = (code: number) => {
	if (typeof process.env.JEST_WORKER_ID === "undefined") {
		// eslint-disable-next-line no-process-exit
		process.exit(code);
	}
};

export default (argv: Array<string>) => {
	try {
		/**
		 * config assembly
		 */

		const userConfig = loadUserConfig();
		const cliConfig = loadCliConfig(argv);
		const config = {
			...DEFAULT_CONFIG,
			...{ json: false },
			...userConfig,
			...cliConfig
		};

		// output the config (--config) but only if we don't want JSON output on the CLI
		if (!config.json && config.config) {
			const json = JSON.stringify(config, null, 2);

			// eslint-disable-next-line no-console
			console.log(`\n\n${json}\n\n`);
		}

		/**
		 * package.json
		 */

		const {
			writePackageScripts,
			readPackageScripts
		} = userPackageScriptContext(
			makePackageFilePath(config.packageFile ?? process.cwd())
		);

		const scripts = readPackageScripts(config.ignoreScripts);

		/**
		 * rule loading
		 */

		const rules = loadRulesFromRuleConfig(
			config.strict,
			config.rules,
			config.customRules
		);

		/**
		 * and go!
		 */
		const [issues, fixedScripts, issuesFixed] = execute(
			rules,
			scripts,
			warning,
			config.fix
		);

		if (config.fix && issuesFixed > 0) {
			writePackageScripts(fixedScripts);
			success(`Fixed ${issuesFixed} issue${issuesFixed > 1 ? "s" : ""}!`);
		}

		if (issues.length - issuesFixed === 0) {
			success("âœ¨  All good");

			dump(config.json);

			processExit(PROCESS_EXIT_OK);
		}

		dump(config.json);
		processExit(PROCESS_EXIT_ERROR);
	} catch (err) {
		error(err);
	}
};
