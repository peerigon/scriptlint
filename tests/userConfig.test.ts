import {DEFAULT_CONFIG} from "../src/constants";
import loadConfig, {sanitizeConfig} from "../src/userConfig";

const validConfig = {
	strict: true,
	fix: false,
	json: false,
	config: false,
	rules: { foo: "bar" },
	ignoreScripts: ["foo"],
	customRules: [
		{
			isObjectRule: false,
			name: "foobar",
			message: "barbaz",
			validate: () => true
		}
	]
};

const invalidConfig = {
	invalid: 3
};

describe("userConfig.ts", () => {
	it("should sanitize configs: empty config => default", () => {
		expect(sanitizeConfig({})).toEqual(DEFAULT_CONFIG);
	});
	it("should sanitize configs: null => default", () => {
		expect(sanitizeConfig(null)).toEqual(DEFAULT_CONFIG);
	});
	it("should sanitize configs: invalid keys", () => {
		expect(() => {
			sanitizeConfig(invalidConfig);
		}).toThrowErrorMatchingSnapshot();
	});
	it("should sanitize configs: valid => valid", () => {
		expect(sanitizeConfig(validConfig)).toEqual(validConfig);
	});

	test("loadConfig()", () => {
		const loaded = loadConfig();

		expect(loaded).toMatchSnapshot();
	});

	test("loadConfig() with config missing", () => {
		const loaded = loadConfig("missing");

		expect(loaded).toBe(DEFAULT_CONFIG);
	});
});

export {};
