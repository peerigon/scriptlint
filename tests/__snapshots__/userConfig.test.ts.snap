// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userConfig.ts loadConfig() 1`] = `
Object {
  "config": false,
  "customRules": Array [],
  "fix": false,
  "ignoreScripts": Array [
    "foobar",
  ],
  "json": true,
  "packageFile": undefined,
  "rules": Object {
    "mandatory-dev": false,
  },
  "strict": true,
}
`;

exports[`userConfig.ts should sanitize configs: invalid keys 1`] = `"unknown config key \\"invalid\\""`;
